File: src/main/java/com/bankingsystem/service/TransactionService.java
package com.bankingsystem.service;

import com.bankingsystem.model.Transaction;

import java.math.BigDecimal;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface TransactionService {

    /**
     * Credits the specified amount to the user's account.
     *
     * @param userId ID of the user to credit
     * @param amount Amount to credit
     * @return Transaction record of the credit
     */
    Transaction credit(Long userId, double amount);

    /**
     * Debits the specified amount from the user's account.
     *
     * @param userId ID of the user to debit
     * @param amount Amount to debit
     * @return Transaction record of the debit
     */
    Transaction debit(Long userId, double amount);

    /**
     * Transfers amount from one user to another.
     *
     * @param fromUserId Sender user ID
     * @param toUserId Receiver user ID
     * @param amount Amount to transfer
     * @return Transaction record of the transfer
     */
    Transaction transfer(Long fromUserId, Long toUserId, double amount);

    /**
     * Retrieves the transaction history of a user.
     *
     * @param userId ID of the user
     * @return List of transactions
     */
    List<Transaction> getTransactionsByUser(Long userId);

    /**
     * Retrieves the current balance of a user.
     *
     * @param userId ID of the user
     * @return BigDecimal balance
     */
    BigDecimal getBalance(Long userId);

    Page<Transaction> getTransactionsByUser(Long userId, Pageable pageable);
}

File: mini-banking-frontend/src/services/transactionService.js
// src/services/transactionService.js
// Transaction-related API helpers.
// Uses the `api` axios instance (src/services/api.js).

import api from './api';

const getTokenHeader = () => {
  const token = localStorage.getItem('token');
  return token ? { Authorization: `Bearer ${token}` } : {};
};

/**
 * Get paginated transactions for a user.
 * Calls: GET /api/transactions/{userId}?page=&size=&sortBy=&sortDir=
 */
export async function getTransactions(userId, { page = 0, size = 10, sortBy = 'transactionDate', sortDir = 'desc' } = {}) {
  try {
    const res = await api.get(`/transactions/${userId}`, {
      headers: getTokenHeader(),
      params: { page, size, sortBy, sortDir },
    });
    return res.data;
  } catch (err) {
    throw err?.response?.data || new Error(err.message || 'Fetch transactions failed');
  }
}

/**
 * Get non-paginated history array endpoint if your backend exposes /history/{userId}
 */
export async function getAllTransactions(userId) {
  try {
    const res = await api.get(`/transactions/history/${userId}`, { headers: getTokenHeader() });
    return res.data;
  } catch (err) {
    throw err?.response?.data || new Error(err.message || 'Fetch transaction history failed');
  }
}

/**
 * Get account balance
 * Calls: GET /api/transactions/balance/{userId}
 */
export async function getBalance(userId) {
  try {
    const res = await api.get(`/transactions/balance/${userId}`, { headers: getTokenHeader() });
    return res.data;
  } catch (err) {
    throw err?.response?.data || new Error(err.message || 'Fetch balance failed');
  }
}

/**
 * Credit account
 * Calls: POST /api/transactions/credit  with body { userId, amount }
 */
export async function credit(userId, amount) {
  try {
    const res = await api.post('/transactions/credit', { userId, amount }, { headers: getTokenHeader() });
    return res.data;
  } catch (err) {
    throw err?.response?.data || new Error(err.message || 'Credit failed');
  }
}

/**
 * Debit account
 * Calls: POST /api/transactions/debit  with body { userId, amount }
 */
export async function debit(userId, amount) {
  try {
    const res = await api.post('/transactions/debit', { userId, amount }, { headers: getTokenHeader() });
    return res.data;
  } catch (err) {
    throw err?.response?.data || new Error(err.message || 'Debit failed');
  }
}

/**
 * Transfer money
 * Calls: POST /api/transactions/transfer  with body { fromUserId, toUserId, amount }
 */
export async function transfer(fromUserId, toUserId, amount) {
  try {
    const res = await api.post('/transactions/transfer', { fromUserId, toUserId, amount }, { headers: getTokenHeader() });
    return res.data;
  } catch (err) {
    throw err?.response?.data || new Error(err.message || 'Transfer failed');
  }
}

export default {
  getTransactions,
  getAllTransactions,
  getBalance,
  credit,
  debit,
  transfer,
};
